### PR TITLE
Adds exception handling to Agent XAPI client

### PR DESCRIPTION
**Pull request by Cerberus98**
RM11721

Adds exception handling and logging to the sessioned() method of the
quark agent. Previously a failure would be swallowed, forcing us to
manually debug where the error was.

### CODE REVIEW

File: quark/agent/xapi.py [hunk 1]
```diff
@@ -108,11 +108,16 @@ def _session(self):
...
         try:
             session = self._session()
             yield session
+        except Exception:
+            LOG.exception("Failed to create a XAPI session")
+            raise
         finally:
-            session.logout()
+            if session:
+                session.logout()
```
Comment (asadoughi) [inline quark/agent/xapi.py]: try/yield/finally. That's a neat pattern.

**Comment by asadoughi** [comment]
:+1: waiting for travis-ci

File: quark/agent/xapi.py [hunk 2]
```diff
@@ -108,11 +108,16 @@ def _session(self):
 
     @contextlib.contextmanager
     def sessioned(self):
+        session = None
         try:
             session = self._session()
             yield session
+        except Exception:
+            LOG.exception("Failed to create a XAPI session")
```
Comment (insequent) [inline quark/agent/xapi.py]: Would it be useful to log the excepton message as well? Not familiar enough with this to know if meaningful messages come from xapi handling.

Comment (Cerberus98) [inline quark/agent/xapi.py]: Doesn't log.exception automatically find and log it?

Comment (Cerberus98) [inline quark/agent/xapi.py]: Just removed the assertRaises clause and tried again, and got this:

<pre>
-------------------- >> begin captured logging << --------------------
quark.agent.xapi: DEBUG: Created new Xapi session
quark.agent.xapi: ERROR: Failed to create a XAPI session
Traceback (most recent call last):
  File "/home/nova/code/python/quark/quark/agent/xapi.py", line 113, in sessioned
    session = self._session()
  File "/home/nova/code/python/quark/.tox/py27/local/lib/python2.7/site-packages/mock.py", line 955, in __call__
    return _mock_self._mock_call(*args, **kwargs)
  File "/home/nova/code/python/quark/.tox/py27/local/lib/python2.7/site-packages/mock.py", line 1010, in _mock_call
    raise effect
Failure: HANDLE_INVALID
--------------------- >> end captured logging << ---------------------
</pre>

**Comment by Cerberus98** [comment]
ugh, need to cover the neutron.conf thinger

File: quark/agent/xapi.py (hunk 2 above)
Comment (insequent) [inline quark/agent/xapi.py]: Yup, I must be out of it. I didn't even notice the `.exception`.

**Comment by insequent** [comment]
neutron.conf fun aside, LGTM :snowman:

**Comment by Cerberus98** [comment]
There, think I have it fixed in this patch. I can yank it out into a separate one if you guys prefer.

**Comment by Cerberus98** [comment]
lollerkates

**Comment by Cerberus98** [comment]
One more time

**Comment by Cerberus98** [comment]
\o/

**Comment by Cerberus98** [comment]
Alright, self merging since you johnnies said it was ok.

**Pull request merged**